fuel-core/src/schema/coin.rs

```diff
@@ -167,4 +178,216 @@ impl CoinQuery {
+
+        #[derive(Debug, Error)]
+        pub enum CoinLookupError {
+            #[error("store error occured")]
+            KvStoreError(KvStoreError),
+            #[error("state error occured")]
+            StateError(state::Error),
+            #[error("enough coins could not be found")]
+            NotEnoughCoins,
+            #[error("not enough inputs")]
+            NotEnoughInputs,
+        }
+
+        impl From<KvStoreError> for CoinLookupError {
+            fn from(e: KvStoreError) -> Self {
+                CoinLookupError::KvStoreError(e)
+            }
+        }
+
+        impl From<state::Error> for CoinLookupError {
+            fn from(e: state::Error) -> Self {
+                CoinLookupError::StateError(e)
+            }
+        }
+
+        let owned_coin_ids: Vec<Vec<Bytes32>> = spend_query
+            .iter()
+            .map(|(color, _)| {
+                db.owned_coins_by_color(owner, *color, None, None)
+                    .try_collect()
+            })
+            .try_collect()?;
+
+        let largest_first =
+            |spend_query: Vec<(FuelTxColor, u64)>| -> Result<Vec<Coin>, CoinLookupError> {
+                let mut coins: Vec<Coin> = vec![];
+
+                for (index, (_color, amount)) in spend_query.iter().enumerate() {
+                    let coin_ids = &owned_coin_ids[index];
+                    let mut collected_amount = 0u64;
+
+                    let mut sorted_coins: Vec<(&Bytes32, CoinModel)> = coin_ids
+                        .iter()
+                        .map(|id| {
+                            Storage::<Bytes32, CoinModel>::get(db, id)
+                                .transpose()
+                                .ok_or(KvStoreError::NotFound)?
+                                .map(|coin| (id, coin.into_owned()))
+                        })
+                        .try_collect()?;
+
+                    sorted_coins.sort_by_key(|coin| coin.1.amount);
+
+                    for (id, coin) in sorted_coins {
+                        // Break if we don't need any more coins
+                        if collected_amount >= *amount {
+                            break;
+                        }
+
+                        // Error if we can't fit more coins
+                        if coins.len() >= max_inputs as usize {
+                            return Err(CoinLookupError::NotEnoughInputs);
+                        }
+
+                        // Add to list
+                        collected_amount += coin.amount;
+                        coins.push(Coin(*id, coin));
+                    }
+
+                    if collected_amount < *amount {
+                        return Err(CoinLookupError::NotEnoughCoins);
+                    }
+                }
+
+                Ok(coins)
+            };
+
+        // An implementation of the method described on: https://iohk.io/en/blog/posts/2018/07/03/self-organisation-in-coin-selection/
+        let random_improve =
```

Comment:
These should be extracted into smaller modules/functions instead of defined in-line.

Reply:
Addressed in this commit: https://github.com/FuelLabs/fuel-core/pull/131/commits/6d0ebefdf204333c87027d102f3f0688f16ba95c